### ISSUE TITLE
Absurdly slow compilation on CPU

### ISSUE THREAD

**Issue by rongcuid** [google/jax#7829]
On a CPU run, JIT compiling a loss function that involves forwarding through a complex model causes Very Slow Compilation. I also suspect that it deadlocks, because the program does not respond to interruption and has to be killed by SIGKILL. I tried #7797, but that is not the problem. I also checked a few closed issues, but those are related to loops, and I use nothing of that. I used vmap in many places, though.

I am unable to provide a smaller example that reproduces this problem, so I will link to my repository instead.

https://github.com/rongcuid/flax-transformer/commit/752a48ff52fc7a404e2dab24e48b2f291af24da6

In `train.py`, there is a commented out `@jax.jit` in `train_step()`. Uncomment it and bug happens.

    $ python train.py --steps 1 --batch-size 1 output

**Comment by shoyer** [comment]
Can you try running `jax.make_jaxpr` on your code? How long is the output from `make_jaxpr`?
https://jax.readthedocs.io/en/latest/jax.html#jax.make_jaxpr

If `make_jaxpr` runs quickly, this indicates that the slowness is likely within the XLA compiler rather than JAX itself. If it's an issue with the XLA CPU backend specifically, it might also help to run your code on a GPU or TPU.

**Comment by rongcuid** [comment]
Hello, both `jax.jit` and `jax.make_jaxpr` returns immediately. Worse is that the `loss_fn` which was taken gradients of could be compiled in reasonable time when I single stepped it. When I tried to run it directly, it seems to get stuck forever.

While I do have access to a GPU, I am trying to debug on a laptop that does not have one.

**Comment by shoyer** [comment]
> Hello, both `jax.jit` and `jax.make_jaxpr` returns immediately

Can you clarify? I thought `jax.jit` was hanging.

**Comment by rongcuid** [comment]
> > Hello, both `jax.jit` and `jax.make_jaxpr` returns immediately
> 
> Can you clarify? I thought calling the function decorated with `jax.jit` was hanging.

Calling `jax.jit(loss_fn)` is instantaneous, as is `make_jaxpr`.  Jit returns a `CompiledFunction` object, but when actually executing the `grad_fn` the program seems to hang. In fact, it does not respond to anything, be it interruption, or even debugger pause.

**Comment by rongcuid** [comment]
Apparently, when I single stepped it in the debugger (with `loss_fn` decorated and `train_step` undecorated), the function can progress after a few minutes of execution. So the `loss_fn` decoration is probably not the cause.

The `train_step` decoration, however, appears to hang when the function _returns_.

**Comment by Chillee** [comment]
@rongcuid To clarify, `jax.jit` and `jax.make_jaxpr` do not do the compilation/jaxpr creation, they simply wrap the function in a new function that does the compilation/jaxpr creation when you call them.

You need to call `jax.make_jaxpr(f)(*args)` to actually create the jaxpr.

Similarly, after you do `f2 = jax.jit(f)`, the first time you call `f2(*args)` is when Jax does the actual compilation.

**Comment by rongcuid** [comment]
I get an enormous output like this:

[output.txt](https://github.com/google/jax/files/7124569/output.txt)

**Comment by shoyer** [comment]
This suggests to me that maybe something in your code _is_ effectively doing a loop to generate lots of code for XLA. It might be worth experimenting with your code to see if commenting out part of it makes things much faster.
